subnet/subnet_contract.go

```diff
@@ -0,0 +1,137 @@
+package subnet
+
+import (
+	"github.com/nspcc-dev/neo-go/pkg/interop"
+	"github.com/nspcc-dev/neo-go/pkg/interop/contract"
+	"github.com/nspcc-dev/neo-go/pkg/interop/native/management"
+	"github.com/nspcc-dev/neo-go/pkg/interop/runtime"
+	"github.com/nspcc-dev/neo-go/pkg/interop/storage"
+	"github.com/nspcc-dev/neofs-contract/common"
+)
+
+const (
+	// ErrInvalidSubnetID is thrown when subnet id is not a slice of 4 bytes.
+	ErrInvalidSubnetID = "invalid subnet ID"
+	// ErrInvalidOwner is thrown when owner has invalid format.
+	ErrInvalidOwner = "invalid owner"
+	// ErrAlreadyExists is thrown when id already exists.
+	ErrAlreadyExists = "subnet id already exists"
+	// ErrNotExist is thrown when id doesn't exist.
+	ErrNotExist = "subnet id doesn't exist"
+
+	ownerPrefix       = 'o'
+	infoPrefix        = 'i'
+	notaryDisabledKey = 'z'
+)
+
+// _deploy function sets up initial list of inner ring public keys.
+func _deploy(data interface{}, isUpdate bool) {
+	if isUpdate {
+		return
+	}
+
+	args := data.(struct {
+		notaryDisabled bool
+	})
+
+	ctx := storage.GetContext()
+	storage.Put(ctx, []byte{notaryDisabledKey}, args.notaryDisabled)
+}
+
+// Put creates new subnet with the specified owner and info.
+func Put(id []byte, ownerKey interop.PublicKey, info []byte) {
+	if len(id) != 4 {
+		panic("put: " + ErrInvalidSubnetID)
+	}
+	if len(ownerKey) != interop.PublicKeyCompressedLen {
+		panic("put: " + ErrInvalidOwner)
+	}
+
+	ctx := storage.GetContext()
+	notaryDisabled := storage.Get(ctx, notaryDisabledKey).(bool)
+	if notaryDisabled {
+		alphabet := common.AlphabetNodes()
+		nodeKey := common.InnerRingInvoker(alphabet)
+		if len(nodeKey) == 0 {
+			if !runtime.CheckWitness(ownerKey) {
+				panic("put: witness check failed")
+			}
+			runtime.Notify("SubnetPut", ownerKey, info)
+			return
+		}
+
+		threshold := len(alphabet)*2/3 + 1
+		id := common.InvokeID([]interface{}{ownerKey, info}, []byte("put"))
+		n := common.Vote(ctx, id, nodeKey)
+		if n < threshold {
+			return
+		}
+
+		common.RemoveVotes(ctx, id)
+	} else {
+		if !runtime.CheckWitness(ownerKey) {
+			panic("put: witness check failed")
+		}
+
+		multiaddr := common.AlphabetAddress()
+		if !runtime.CheckWitness(multiaddr) {
+			panic("put: alphabet witness check failed")
+		}
+	}
+
+	stKey := append([]byte{ownerPrefix}, id...)
+	if storage.Get(ctx, stKey) != nil {
+		panic("put: " + ErrAlreadyExists)
+	}
```

Comment:
Put this check before `if notaryDisable` condition. Early checks before `runtime.Notify` are appreciated by the clients and it saves a bit of wasted GAS in notary disabled environment.